packages/runtime/container-runtime/src/channelCollection.ts

```diff
@@ -819,84 +819,121 @@ export class ChannelCollection implements IFluidDataStoreChannel, IDisposable {
-				return;
-			case ContainerMessageType.FluidDataStoreOp: {
-				const envelope = message.contents as IEnvelope;
-				const innerContents = envelope.contents as FluidDataStoreMessage;
-				const transformed = {
-					...message,
-					type: innerContents.type,
-					contents: innerContents.content,
-				};
-
-				this.processChannelOp(envelope.address, transformed, local, localMessageMetadata);
-
-				// Notify GC of any outbound references that were added by this op.
-				detectOutboundReferences(
-					envelope.address,
-					transformed.contents,
-					(fromPath: string, toPath: string) =>
-						this.parentContext.addedGCOutboundRoute(fromPath, toPath, message.timestamp),
-				);
-				break;
-			}
-			default:
-				assert(false, 0x8e9 /* unreached */);
-		}
+		throw new Error("process should not be called.");
 	}
 
-	protected processChannelOp(
-		address: string,
-		message: ISequencedDocumentMessage,
+	/**
+	 * Process channel type messages. The messages here are contiguous channel type messages in a batch. Bunch
+	 * of contiguous messages for a data store should be sent to it together.
+	 * @param messagesWithMetadata - A bunch of contiguous channel type messages and their metadata.
+	 * @param local - true if the messages were originally generated by the client receiving it.
+	 */
+	public processChannelMessages(
+		messagesWithMetadata: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata: unknown;
+		}[],
 		local: boolean,
-		localMessageMetadata: unknown,
-	) {
-		const context = this.contexts.get(address);
-
-		// If the data store has been deleted, log an error and ignore this message. This helps prevent document
-		// corruption in case a deleted data store accidentally submitted an op.
-		if (this.checkAndLogIfDeleted(address, context, "Changed", "processFluidDataStoreOp")) {
+	): void {
+		if (messagesWithMetadata.length === 0) {
 			return;
 		}
 
-		if (context === undefined) {
-			// Former assert 0x162
-			throw DataProcessingError.create(
-				"No context for op",
-				"processFluidDataStoreOp",
-				message,
-				{
-					local,
-					messageDetails: JSON.stringify({
-						type: message.type,
-						contentType: typeof message.contents,
-					}),
-					...tagCodeArtifacts({ address }),
-				},
-			);
-		}
+		/**
+		 * Bunch contiguous messages for the same data store and send them together.
+		 * This is an optimization mainly for DDSes, where it can process a bunch of ops together. DDSes
+		 * like merge tree or shared tree can process ops more efficiently when they are bunched together.
+		 */
+		let previousAddress: string | undefined;
+		let contextMessages: {
+			message: ISequencedDocumentMessage;
+			localOpMetadata: unknown;
+		}[] = [];
+
+		messagesWithMetadata.forEach((messageWithMetadata) => {
+			const { message, localOpMetadata } = messageWithMetadata;
+			const envelope = message.contents as IEnvelope;
+			const address = envelope.address;
```

Comment:
In other places you check that address !== undefined. This might not be required (as the rest of the logic would fail anyway), but it might not be clear to the reader that's the starting point